### ISSUE TITLE
Bug: Can't unignore file in flat config

### ISSUE THREAD

**Issue by fisker** [eslint/eslint#17964]
### Environment

Node version: v18.18.0
npm version: v9.4.2
Local ESLint version: v8.56.0 (Currently used)
Global ESLint version: Not found
Operating System: linux 5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```
export default [
  {
    ignores: ['files/**', '!files/should-be-linted.js'],
  },
  {
    rules: {
      'no-var': 'error',
    },
  },
];

```
</details>

```js
var a = 1;
```

### What did you expect to happen?

Should report error for `files/should-be-linted.js` file.

### What actually happened?

```
You are linting "files", but all of the files matching the glob pattern "files" are ignored.
```

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/stackblitz-starters-48wnlw?description=Starter%20project%20for%20Node.js,%20a%20JavaScript%20runtime%20built%20on%20Chrome%27s%20V8%20JavaScript%20engine&file=package.json,eslint.config.js,files%2Fshould-be-linted.js&title=node.new%20Starter

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
> ignores: ['files/**', '!files/should-be-linted.js'],

This works as intended because, in flat config, these are minimatch patterns, and per the minimatch rules `files/**` also matches `files/` directory (https://github.com/isaacs/minimatch/issues/179), which is not unignored by a subsequent pattern so everything under it is still ignored.

The following patterns would ignore everything under `foo/` except for `files/should-be-linted.js`:

```
ignores: ['files/*', '!files/should-be-linted.js']
```

I think we should explain this better in the documentation and migration guide.

**Comment by fisker** [comment]
Question: How to unignore file in `files/`?

**Comment by mdjermanovic** [comment]
> Question: How to unignore file in `files/`? I can't figure out, it blocks me switch to flat config in Prettier [prettier/prettier#15869](https://github.com/prettier/prettier/pull/15869).

The problem is that the following doesn't unignore `jsfmt.spec.js` files?

```
tests/format/**/*
!tests/format/**/jsfmt.spec.js
```

It's most likely because `tests/format/**/*` matches directories where `jsfmt.spec.js` files are, can you try with:

```
`tests/format/**/*.js`
!tests/format/**/jsfmt.spec.js
```

**Comment by fisker** [comment]
Thanks, it seems works! I'm going to use

```
tests/format/**/*.*
!tests/format/**/jsfmt.spec.js
```

**Comment by fisker** [comment]
I have to say it's really odd.. 

I guess, my patterns won't work if directory contains `.` or the file wan't be ignored don't contains `.` (eg: bin files without extensions).

Another thing is that it seems we treat patterns starts with `/` as absolute path, which is also different as before, that's why I removed all leading `/` in https://github.com/prettier/prettier/pull/15869/files

**Comment by mdjermanovic** [comment]
> I guess, my patterns won't work if directory contains `.` or the file want be unignored doesn't contains `.` (eg: bin files without extensions).

I think this is the same problem as with .gitignore patterns, and the solution would be similar:

```
# first, ignore everything (files and directories)
tests/format/**/*

# then, uningore all directories
!tests/format/**/*/

# then, unignore specific files
!tests/format/**/jsfmt.spec.js

```

That said, some behavior of flat config ignore patterns does look odd, I'll double-check whether everything works as intended.

> Another thing is that it seems we treat patterns starts with `/` as absolute path, which is also different as before, that's why I removed all leading `/` in https://github.com/prettier/prettier/pull/15869/files

This is intended. In flat config, patterns should not start with `/` .

This is different from .gitignore, where `/a.js` pattern matches only top-level `a.js`, while `a.js` pattern matches `a.js` at any level. In flat config, `a.js` pattern matches only top-level `a.js`, `**/a.js` should be used to match at any level, while `/a.js` is basically invalid. We should document this as well.

**Comment by fisker** [comment]
```
tests/format/**/*
!tests/format/**/*/
!tests/format/**/jsfmt.spec.js
```

Doesn't work as I tested, it will unignore all files.

**Comment by mdjermanovic** [comment]
> ```
> tests/format/**/*
> !tests/format/**/*/
> !tests/format/**/jsfmt.spec.js
> ```
> 
> 
>     
> 
>   
> 
> Doesn't work as I tested, it will unignore all files.

I'd expect this to ignore everything under `tests/format/` except for `jsfmt.spec.js`. I'll check why it doesn't work that way.

**Comment by mdjermanovic** [comment]
> > ```
> > tests/format/**/*
> > !tests/format/**/*/
> > !tests/format/**/jsfmt.spec.js
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Doesn't work as I tested, it will unignore all files.
> 
> I'd expect this to ignore everything under `tests/format/` except for `jsfmt.spec.js` files. I'll check why it doesn't work that way.

Looks like a bug to me, I'm working on this.

**Comment by nzakas** [comment]
@mdjermanovic https://github.com/eslint/eslint/pull/17980 upgrades `config-array`, but you may want to add tests to cover this issue.

**Comment by mdjermanovic** [comment]
> @mdjermanovic #17980 upgrades `config-array`, but you may want to add tests to cover this issue.

Yes, I'll add test cases.

**Comment by mdjermanovic** [comment]
Tests: https://github.com/eslint/eslint/pull/18018

I'm working on the docs now.
